### ISSUE TITLE
Casting pointers can hide inttoptr, causing Points-to-graph to lack edges.

### ISSUE THREAD

**Issue by haved** [phate/jlm#252]
LLVM IR instructions like `inttoptr` and `ptrtoint` make it very clear that integer-pointer-conversions are happening, and functions like `Steensgaard::AnalyzeBits2Ptr` will correctly mark outputs as pointing to unknown or external memory.

We can however "trick" the compiler by never doing any casts between integers and pointers, and only doing pointer-to-pointer casts. Detecting this from the LLVM IR is much harder, since the `ptr` type has no info about what it points at.

## Reproduction
Take the following C program:
```C
uint32_t ptrCastTricks(uintptr_t external_int) {
    uint32_t data = 0;

    uint32_t* data_ptr = &data;
    uint32_t** danger = &data_ptr;

    * (uintptr_t*)danger = external_int;
    return *data_ptr;
}
```

which effectively does exactly the same as
```C
uint32_t notSneaky(uintptr_t external_int) {
    uint32_t data = 0;

    uint32_t* data_ptr = &data;
    data_ptr = (uint32_t*) external_int;

    return *data_ptr;
}
```

<details>
<summary>
This will result in the following unoptimized LLVM IR
</summary>

```
define dso_local i32 @ptrCastTricks(i64 noundef %0) #0 {
  %2 = alloca i64, align 8
  %3 = alloca i32, align 4
  %4 = alloca ptr, align 8
  %5 = alloca ptr, align 8
  store i64 %0, ptr %2, align 8
  store i32 0, ptr %3, align 4
  store ptr %3, ptr %4, align 8
  store ptr %4, ptr %5, align 8
  %6 = load i64, ptr %2, align 8
  %7 = load ptr, ptr %5, align 8
  store i64 %6, ptr %7, align 8
  %8 = load ptr, ptr %4, align 8
  %9 = load i32, ptr %8, align 4
  ret i32 %9
}
```
</details>

Note how this program has no `inttoptr` despite in practice converting the `i64 %0`  to a `ptr`, by storing the `i64`, and reading it back as a `ptr`.

<details>
<summary>

The output of `jlm-opt --AASteensgaardAgnostic` becomes
</summary>

![image](https://github.com/phate/jlm/assets/3748845/b0f83549-a6cc-4fea-9524-e54bdc8bd2cf)
</details>

Note how the program has no `BITS2PTR` or `BITCAST`, only stores and loads. Thanks to the pointer casting, we are able to perform `data_ptr = (uint32_t*) external_int` without any explicit conversion.

The Points-to-graph produced by Steensgaard becomes this:
![image](https://github.com/phate/jlm/assets/3748845/b21602ec-5c98-4d57-8c58-22868ebe6cd7)

## Expected output
Since we effectively override `data_ptr = (uint32_t*)external_int`, we expect the `ALLOCA[ptr]` node for `data_ptr` to point to UnkownMemory and ExternalMemory, but it does not.

<details>
<summary>

Here is a picture of the Points-to-graph from the `notSneaky()` version:
</summary>

![image](https://github.com/phate/jlm/assets/3748845/0fc2c8c0-9209-4b07-bc4b-3b9f2e5e6229)

The total number of edges is a lot higher, probably due to a combination of the facts that integers can point to a lot of things, and the unification done by Steensgaard.
</details>

## Potential issues this causes
The AA pass has done a great job of  rerouting memory state edges, but the final `LOAD` in the graph will believe that its input only ever points to `data`, which causes its memory state edge to only depend on the previous store to `data`, even if other things had happened to `*external_int` in the meantime.

## Suggested fix
One possible way to fix this (other than to track points-to-info about all integers), is to make stores of integers update the points-to-edges of all memory objects it might override, to point to Unknown and External. It would look something like this:
![image](https://github.com/phate/jlm/assets/3748845/8a7592d4-e52b-491e-b1e9-4fddca044ae0)

This would be a relatively simple change in `Steensgaard::AnalyzeStore`, which currently early-returns if the value is not a pointer. The downside is that all ALLOCA[non-ptr] memory nodes start pointing to Unknown and External, even if it never gets read as a pointer.

To optimize this slightly, we can say that all ALLOCA[non-ptr] memory nodes implicictly point to unknown, and thus only need to add edges when integers are stored into ALLOCA[ptr] nodes. Any pointer value that is possibly read from an ALLOCA[non-ptr] will then need to explicitly add the edge to unknown.

`Steensgaard.cpp` luckily uses flags for these special nodes, which hopefully helps to prevent unification of the entire universe into a single DisjointLocationSet.

I'm still figuring out the exact semantics of the `UnknownNode`, `ExternalNode`, and memory objects escaping, so there might be more efficient ways of encoding and handling this.

**Comment by haved** [comment]
I'm working on a proposal for full description of the semantics of the Points-to-graph, and the operations in RVSDG's LLVM dialect, that will work for both Steensgaard and Andersen (anything {flow,context,field}-insensitive, really), and is also inter-procedural.

I'm coming across several sneaky edge-cases, so I will try to document them all.

One of them is the "inverse" of this issue, where a load through a double pointer casted to `(uintptr_t*)` can leak the address of a memory object as a `uintptr_t`, and as such let the address escape the module.

**Comment by haved** [comment]
To keep discussion in the related issue:
When I wrote this issue, I did not consider a case like the following:
```c
struct {
    int* array;
    int length;
} my_list;

my_list.length = 20;
my_list.array = calloc(20, sizeof int);
```

This code would result in two store operations to the same alloca, at different offsets.
One of the stores is an `i32`, while the other is a `ptr` store.

Since the AA is field-insensitive, we don't really distinguish between writes to the different fields.
With what I proposed above, the entire alloca memory object would point to unknown, due to storing an integer, that in no way overrides the pointer. This makes analysis a lot weaker for fat pointers, so it would be nice if a more precise solution could be found. Making the analysis field-sensitive would be one such solution, but that would make it bigger and more complex, and in the case of Steensgaard remove the nice almost-linear size/runtime property.

Due to C and C++'s type based aliasing rules (strict aliasing), a program is not allowed to access the same bytes through both an `int*` and a `ptr*`, where `ptr` is any pointer type. It is however legal to cast any pointer into a `char*`, and use that to access the memory. This is special for `char`, (and `std::byte` in C++) and any user defined type with `__attribute__((may_alias))`.

In LLVM IR, this can be observed by stores to `int*` getting TBAA tag "int", while stores to a `char*` getting the TBAA tag "omnipotent char", which is the parent TBAA node for both "int" and "any pointer". If the TBAA information is kept, a rule to not add the pointer to unknown unless the written value is of a type that may alias pointers, would help to avoid extra edges to the unknown node.

Strict aliasing is still a bit too strict to some people, so compilers add the `-fno-strict-aliasing` option, making any uses of casted pointers legal. Compilers also explicitly allow "type-punning" through the use of different union fields, even when `-fstrict-aliasing` is enabled. I'm not sure if/how one should go about supporting that, but I just mention it here.
